task-impls/src/consensus.rs

```diff
@@ -248,29 +284,100 @@ where
+            // // error!("in vote if able for proposal view {:?}", proposal.view_number);
+
+            if proposal.justify_qc.is_genesis() && proposal.view_number == ViewNumber::new(1) {
+                // error!("Proposal is genesis!");
+
+                let view = TYPES::Time::new(*proposal.view_number);
+                let vote_token = self.quorum_exchange.make_vote_token(view);
+
+                match vote_token {
+                    Err(e) => {
+                        error!("Failed to generate vote token for {:?} {:?}", view, e);
+                    }
+                    Ok(None) => {
+                        info!("We were not chosen for consensus committee on {:?}", view);
+                    }
+                    Ok(Some(vote_token)) => {
+                        let justify_qc = proposal.justify_qc.clone();
+                        let parent = if justify_qc.is_genesis() {
+                            self.genesis_leaf().await
+                        } else {
+                            self.consensus
+                                .read()
+                                .await
+                                .saved_leaves
+                                .get(&justify_qc.leaf_commitment())
+                                .cloned()
+                        };
+
+                        // Justify qc's leaf commitment is not the same as the parent's leaf commitment, but it should be (in this case)
+                        let Some(parent) = parent else {
+                            error!("Proposal's parent missing from storage with commitment: {:?}", justify_qc.leaf_commitment());
+                            return false;
+                        };
+                        let parent_commitment = parent.commit();
+
+                        let leaf: SequencingLeaf<_> = SequencingLeaf {
+                            view_number: view,
+                            height: proposal.height,
+                            justify_qc: proposal.justify_qc.clone(),
+                            parent_commitment,
+                            deltas: Right(proposal.block_commitment),
+                            rejected: Vec::new(),
+                            timestamp: time::OffsetDateTime::now_utc().unix_timestamp_nanos(),
+                            proposer_id: self.quorum_exchange.get_leader(view).to_bytes(),
+                        };
+
+                        let message: GeneralConsensusMessage<TYPES, I>;
+                        message = self.quorum_exchange.create_yes_message(
+                            proposal.justify_qc.commit(),
+                            leaf.commit(),
+                            view,
+                            vote_token,
+                        );
+
+                        if let GeneralConsensusMessage::Vote(vote) = message {
+                            error!("Sending vote to next leader {:?}", vote.current_view());
+                            self.event_stream
```

Comment:
I think we can't vote here because we haven't checked the DA cert

Reply:
Yes, you are right!  I'll add a comment for now and we can come back to it later.